File: A_PS/pythonPs/P_호텔 대실.py
import heapq
book_time = [["00:01","00:10"],["00:19","00:29"]]
def solution(book_time):
    answer = 0
    result = []
    for i in range(len(book_time)):
        for j in range(len(book_time[i])):
            h,m = map(int, book_time[i][j].split(":"))
            st = h*60 + m
            book_time[i][j] = st
    
    book_time.sort()
    heapq.heappush(result, book_time[0][1])
    
    
    for i in range(1, len(book_time)):
        if result[0] + 10 > book_time[i][0]:
            heapq.heappush(result,book_time[i][1])
        else:
            heapq.heappop(result)
            heapq.heappush(result, book_time[i][1])
    answer = len(result)   
    return answer

print(solution(book_time))

File: A_PS/javaPs/P_호텔대실.java
package javaPs;

import java.util.*;
public class P_호텔대실 {
    class Solution {
        public int solution(String[][] book_time) {
            int answer = 0;
            int[][] intBookTime = new int[book_time.length][2];

            for(int i=0; i<book_time.length; i++) {
                intBookTime[i][0] = Integer.parseInt(book_time[i][0].replace(":", ""));
                intBookTime[i][1] = Integer.parseInt(book_time[i][1].replace(":", ""));

                intBookTime[i][1] += 10;

                if (intBookTime[i][1] % 100 >= 60) intBookTime[i][1] += 40;
            }

            Arrays.sort(intBookTime, (a, b) -> {
                if (a[0] > b[0]) return 1;
                else if (a[0] < b[0]) return -1;
                else {
                    if(a[1] > b[1]) return 1;
                    else return -1;
                }
            });

            PriorityQueue<Integer> q = new PriorityQueue<>();

            q.offer(intBookTime[0][1]);

            for(int i=1; i<intBookTime.length; i++) {
                if(q.peek() > intBookTime[i][0]) q.offer(intBookTime[i][1]);
                else {
                    q.poll();
                    q.offer(intBookTime[i][1]);
                }
            }
            return q.size();
        }
    }
}
